moon/controller_excavator_round2.py

```diff
@@ -1,18 +1,207 @@
 """
   Space Robotics Challenge 2
 """
-from moon.controller import SpaceRoboticsChallenge
+from scipy import spatial
+import numpy as np
+import math
+from datetime import timedelta
+import traceback
 
+from osgar.bus import BusShutdownException
+
+from moon.controller import SpaceRoboticsChallenge, VirtualBumperException, LidarCollisionException, LidarCollisionMonitor
+from osgar.lib.virtual_bumper import VirtualBumper
+from osgar.lib.quaternion import euler_zyx
+
+DIG_SAMPLE_COUNT = 12
+DIG_GOOD_LOCATION_MASS = 10
 
 class SpaceRoboticsChallengeExcavatorRound2(SpaceRoboticsChallenge):
     def __init__(self, config, bus):
         super().__init__(config, bus)
+        bus.register("bucket_dig", "bucket_drop")
+        self.volatile_dug_up = None
+        self.mount_angle = None
+
+    def on_bucket_info(self, bucket_status):
+        self.volatile_dug_up = bucket_status
+
+    def on_joint_position(self, data):
+        super().on_joint_position(data)
+        self.mount_angle = data[self.joint_name.index(b'mount_joint')]
 
     def run(self):
 
-        vol_list = self.send_request('get_volatile_locations').decode('ascii')
-        print (self.time, "main-excavator-round2: Volatiles: %s" % vol_list)
+        def pose_distance(pose1, pose2):
+            return math.hypot(pose1[0] - pose2[0], pose1[1] - pose2[1])
+
+
+        try:
+            print('Wait for definition of last_position and yaw')
```

Comment:
I think you have some "wait for init" function, but it is maybe in some pending PR

Reply:
yes, that's part of the pending feature/moon-routerdealer PR